token/program-2022/src/processor.rs

```diff
@@ -7602,4 +7677,249 @@ mod tests {
+        let excess_lamports = 4_000_000_000_000;
+        multisig_account.lamports = excess_lamports + multisig_minimum_balance();
+        let mut signer_keys = [Pubkey::default(); MAX_SIGNERS];
+
+        for signer_key in signer_keys.iter_mut().take(MAX_SIGNERS) {
+            *signer_key = Pubkey::new_unique();
+        }
+        let signer_refs: Vec<&Pubkey> = signer_keys.iter().collect();
+        let mut signer_lamports = 0;
+        let mut signer_data = vec![];
+        let mut signers: Vec<AccountInfo<'_>> = vec![
+            AccountInfo::new(
+                &destination_key,
+                true,
+                false,
+                &mut signer_lamports,
+                &mut signer_data,
+                &program_id,
+                false,
+                Epoch::default(),
+            );
+            MAX_SIGNERS + 1
+        ];
+        for (signer, key) in signers.iter_mut().zip(&signer_keys) {
+            signer.key = key;
+        }
+
+        let mut multisig =
+            Multisig::unpack_unchecked(&vec![0; Multisig::get_packed_len()]).unwrap();
+        multisig.m = MAX_SIGNERS as u8;
+        multisig.n = MAX_SIGNERS as u8;
+        multisig.signers = signer_keys;
+        multisig.is_initialized = true;
+        Multisig::pack(multisig, &mut multisig_account.data).unwrap();
+
+        let multisig_info: AccountInfo = (&multisig_key, true, &mut multisig_account).into();
+
+        let mut signers_infos = vec![
+            multisig_info.clone(),
+            destination_info.clone(),
+            multisig_info.clone(),
+        ];
+        signers_infos.extend(signers);
+        do_process_instruction_dups(
+            withdraw_excess_lamports(
+                &program_id,
+                &multisig_key,
+                &destination_key,
+                &multisig_key,
+                &signer_refs,
+            )
+            .unwrap(),
+            signers_infos,
+        )
+        .unwrap();
+
+        assert_eq!(destination_info.lamports(), excess_lamports);
+    }
+
+    #[test]
+    #[serial]
+    fn test_withdraw_excess_lamports_from_account() {
+        let excess_lamports = 4_000_000_000_000;
+
+        let program_id = crate::id();
+        let account_key = Pubkey::new_unique();
+        let mut account_account = SolanaAccount::new(
+            excess_lamports + account_minimum_balance(),
+            Account::get_packed_len(),
+            &program_id,
+        );
+
+        let system_program_id = system_program::id();
+        let mut owner_lamports = 0;
+        let mut owner_data = vec![];
+        let owner_key = Pubkey::new_unique();
+        let mut owner_account = SolanaAccount::new(0, 0, &system_program::id());
+        let owner_info = AccountInfo::new(
+            &owner_key,
+            true,
+            false,
+            &mut owner_lamports,
+            &mut owner_data,
+            &system_program_id,
+            false,
+            Epoch::default(),
+        );
```

Comment:
To test the case that I was worried about, could you have `account_account` owned by `account_key` rather than by `owner_key`? You'll probably need to use `process_instruction_dups` for `initialize_account` too

Reply:
Sure, i'll get on it.

Reply:
Could you take a look? No matter what i put in as the authority the process execute successfully... I can't see why

Reply:
You're passing in the correct `AccountInfo`, which is all that really matters in this weird test environment, and not the accounts in the instruction. Here's the patch:

```
diff --git a/token/program-2022/src/processor.rs b/token/program-2022/src/processor.rs
index bdc4556d..49d69954 100644
--- a/token/program-2022/src/processor.rs
+++ b/token/program-2022/src/processor.rs
@@ -7780,20 +7780,6 @@ mod tests {
         );

         let system_program_id = system_program::id();
-        // let mut owner_lamports = 0;
-        // let mut owner_data = vec![];
-        let owner_key = Pubkey::new_unique();
-        // let mut owner_account = SolanaAccount::new(0, 0, &system_program::id());
-        // let owner_info = AccountInfo::new(
-        //     &owner_key,
-        //     true,
-        //     false,
-        //     &mut owner_lamports,
-        //     &mut owner_data,
-        //     &system_program_id,
-        //     false,
-        //     Epoch::default(),
-        // );

         let mut destination_lamports = 0;
         let mut destination_data = vec![];
@@ -7814,7 +7800,7 @@ mod tests {

         let mut rent_sysvar = rent_sysvar();
         do_process_instruction(
-            initialize_mint(&program_id, &mint_key, &owner_key, None, 2).unwrap(),
+            initialize_mint(&program_id, &mint_key, &mint_key, None, 2).unwrap(),
             vec![&mut mint_account, &mut rent_sysvar],
         )
         .unwrap();
@@ -7864,20 +7850,7 @@ mod tests {
         let excess_lamports = 4_000_000_000_000;

         let program_id = crate::id();
-        let mut owner_lamports = 0;
-        let mut owner_data = vec![];
         let system_program_id = system_program::id();
-        let owner_key = Pubkey::new_unique();
-        let owner_info = AccountInfo::new(
-            &owner_key,
-            true,
-            false,
-            &mut owner_lamports,
-            &mut owner_data,
-            &system_program_id,
-            false,
-            Epoch::default(),
-        );

         let mut destination_lamports = 0;
         let mut destination_data = vec![];
@@ -7901,7 +7874,7 @@ mod tests {
         let mut rent_sysvar = rent_sysvar();

         do_process_instruction(
-            initialize_mint(&program_id, &mint_key, &owner_key, None, 2).unwrap(),
+            initialize_mint(&program_id, &mint_key, &mint_key, None, 2).unwrap(),
             vec![&mut mint_account, &mut rent_sysvar],
         )
         .unwrap();
@@ -7909,12 +7882,12 @@ mod tests {
         let account_info: AccountInfo = (&mint_key, true, &mut mint_account).into();

         do_process_instruction_dups(
-            withdraw_excess_lamports(&program_id, &mint_key, &destination_key, &program_id, &[])
+            withdraw_excess_lamports(&program_id, &mint_key, &destination_key, &mint_key, &[])
                 .unwrap(),
             vec![
                 account_info.clone(),
                 destination_info.clone(),
-                owner_info.clone(),
+                account_info.clone(),
             ],
         )
         .unwrap();
```

If you pass in something other than `account_info` in that last changed line, it'll error properly

Reply:
I see now, thanks. It's been some time since i last looked at this, so i forgot the details.